[UI] Display system partitions in node page

ChengYanJin opened the pull request:
**Component**: ui, node

**Context**:  We would like to add a tab in the Node page to display the information of system partitions.

**Summary**:

Update: 
Will address hiding the warning alert in the next PR.


**Acceptance criteria**: 
![image](https://user-images.githubusercontent.com/18453133/105346061-5b293e00-5be5-11eb-9f52-bba21f773811.png)


![image](https://user-images.githubusercontent.com/18453133/105345936-2f0dbd00-5be5-11eb-9bdc-fc2990f764e9.png)


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2932 
Closes: #2173 
<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

ChengYanJin committed:
ui/nodes: Add system devices tab

Refs: #2932

ChengYanJin committed:
ui/app: Add QueryClientProvider to <App/>

Use the QueryClientProvider component to connect and provide a
QueryClient to the application

Refs: #2932

ChengYanJin committed:
ui/translation: Add no system parition translations

Refs: #2932

ChengYanJin committed:
ui/constants: Export the node filesystem alertsname

Refs: #2932

ChengYanJin committed:
ui/utils: Add getNodePartitionsTableData function

To return the data for system partition table

Refs: #2932

ChengYanJin committed:
ui/node: Add Partition tab to the Node page

Refs: #2932

ChengYanJin committed:
ui/node: Update the node list selection

To make sure when switch the node we will still in Partition Tab

Refs: #2932

ChengYanJin committed:
ui/nodes: Add Node Partition Tab

Refs: #2932

ChengYanJin committed:
ui/node: add node partition table

Refs: #2932

ChengYanJin committed:
ui/service: QueryNodeFSAvail that use by react query

Refs: #2932

ChengYanJin committed:
ui/style: Move all the style component to a dedicate file

Refs: #2932

ChengYanJin committed:
ui/typing: Add typing for the result return by prometheus query

Refs: #2932

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

ChengYanJin committed:
ui/package: Introduce React-query, React testing library.

Bump the version of styeld-component and core-ui

Refs: #2932

JBWatenbergScality commented on ui/src/services/NodeVolumesUtils.js:
```diff
@@ -20,13 +21,17 @@ import {
-  persistentVolumes,
+  volumeStatus: string,
+  volumeName: string,
+  persistentVolumes: Array<Object>,
```
You may want to explicitly type `persistentVolumes` instead of using the generic `Object` type. 
It would be of type `V1PersistentVolume` from kubernetes client.

JBWatenbergScality replied:
Also you can add a return type definition `): boolean` after the definition of this function in order to enforce return type

JBWatenbergScality commented on ui/src/services/NodeVolumesUtils.js:
```diff
@@ -20,13 +21,17 @@ import {
-  volumeStatus,
-  volumeName,
-  persistentVolumes,
+  volumeStatus: string,
```
volumeStatus type can be constrained to `"Unknonw" | "Pending" | "Terminating" | "Failed" | "Ready"`, there may have an Enum or an union type already representing this status exposed by kubernetes client

JBWatenbergScality commented on ui/src/services/NodeVolumesUtils.js:
```diff
@@ -78,11 +83,11 @@ export const isVolumeDeletable = (
 // Returns
 //     The computed global status of the volume.
-export const computeVolumeGlobalStatus = (name, status) => {
+export const computeVolumeGlobalStatus = (name: string, status: Object) => {
```
`Object` type should generaly be avoided because it is too generic as it match any kind of object. In the context of this function you expect status to be of type `V1PersistentVolumeClaimStatus`

JBWatenbergScality replied:
You can also enforce return type to `"Unknown" | "Ready" | "Failed" | "Pending" | "Terminating"`. You may want to export this type as it is used in several places

ChengYanJin replied:
The status actually is the `Volume` status, I don't think it's exactly the same `V1PersistentVolumeClaimStatus`

JBWatenbergScality commented on ui/src/services/NodeVolumesUtils.js:
```diff
@@ -125,11 +130,11 @@ export const computeVolumeGlobalStatus = (name, status) => {
 // Returns
 //     a tuple (error code, error message).
-export const volumeGetError = (status) => {
+export const volumeGetError = (status: Object) => {
```
status is actually of type `V1PersistentVolumeClaimStatus` and return type would be `[string, string]`

JBWatenbergScality commented on ui/src/services/NodeVolumesUtils.js:
```diff
@@ -315,7 +320,7 @@ export const formatVolumeCreationData = (newVolumes) => {
  * const formatedVolumeName = formatBatchName(name, 1)
  */
-export const formatBatchName = (name, index) => {
+export const formatBatchName = (name: string, index: number) => {
```
You can add a type annotation for the return type of your function : `(name: string, index: number): string => {`

JBWatenbergScality commented on ui/src/services/NodeVolumesUtils.js:
```diff
@@ -324,3 +329,67 @@ export const formatBatchName = (name, index) => {
+};
+
+export const getNodePartitionsTableData = (
+  avails: Array<Object>,
```
`Object` type should generally be avoided because it is too generic as it match any kind of object.

JBWatenbergScality commented on ui/src/services/NodeVolumesUtils.js:
```diff
@@ -324,3 +329,67 @@ export const formatBatchName = (name, index) => {
+      (partition) => partition.partitionPath === size.metric.mountpoint,
+    );
+    if (partition) {
+      partition.usage = Math.round(
```
When possible I feel it is preferable to avoid mutating objects because it generally leads to sneaky bugs when refactoring and as long as the code base enlarges. My preferable way to define partition usage and size here would be to do it directly in the map across `avails` array.

For eg : 

```javascript
const partitions = avails.map((avail) => {
    const {usage, size} = computePartitionUsageAndSize(partitionPath, partitions, sizes);

    return {
      partitionPath: avail.metric.mountpoint,
      availBytes: avail.value[1],
      device: avail.metric.device,
      usage,
      health: 'health',
      size,
    };
  });
```
and a similar approach could be used for `health` computing

JBWatenbergScality commented on ui/src/services/alertmanager/api.js:
```diff
@@ -1,11 +1,14 @@
+// @flow
```
You may want to cherry pick this commit https://github.com/scality/metalk8s/pull/3002/commits/eea1b97b0afb024657e2203a1ebbdca9f31f77fc here or we should find out a way to get https://github.com/scality/metalk8s/pull/3002 merged

JBWatenbergScality commented on ui/src/components/NodePagePartitionTab.js:
```diff
@@ -0,0 +1,55 @@
+  padding: ${padding.large} ${padding.base} 0 ${padding.larger};
+`;
+
+const NodePagePartitionTab = (props: Object) => {
```
`Object` should be replaced with `{instanceIP: string}`

JBWatenbergScality commented on ui/src/components/NodePagePartitionTab.js:
```diff
@@ -0,0 +1,55 @@
+  const { instanceIP } = props;
+
+  // To redirect to the right Node(Detailed) dashboard in Grafana
+  const config = useSelector((state) => state.config);
```
Once https://github.com/scality/metalk8s/pull/3002e is merged you could use useTypedSelector here

JBWatenbergScality commented on ui/src/components/NodePagePartitionTab.js:
```diff
@@ -0,0 +1,55 @@
+          onClick={() => {}}
+          icon={<i className="fas fa-external-link-alt" />}
+          size={'small'}
+          // We can't redirect to the Node(detailed) Filesystem Detail catagory.
```
Some misspelling here `category`

JBWatenbergScality commented on ui/src/components/NodePartitionTable.js:
```diff
@@ -0,0 +1,203 @@
+
+const NodePartitionTable = ({ instanceIP }: { instanceIP: string }) => {
+  // Access the current theme outside of styled components when working with React Hooks.
+  const themeContext = useContext(ThemeContext);
```
You can also use `useTheme` from styled-component, which basically do the same thing but abstract away the ThemeContext which should remain an implementation detail of styled-component

JBWatenbergScality commented on ui/src/components/NodePartitionTable.js:
```diff
@@ -0,0 +1,203 @@
+    data,
+  });
+
+  return (
```
You may want to display a specific state when either of nodeFSAvailQuery, nodeFSSizeQuery and alertsNodeFS status is loading or in error

JBWatenbergScality commented on ui/src/components/NodePartitionTable.js:
```diff
@@ -0,0 +1,203 @@
+      nodeFSSizeQuery.data.data.result,
+      alertsNodeFS.data,
+    );
+  }
```
Maybe this could also be refactored to something similar to :

```js
const {data: devices, status} = useQuery(
    ['nodeDevices', instanceIP],
    useCallback(() => 
         Promise
              .all([queryNodeFSAvail(instanceIP), queryNodeFSSize(instanceIP), getAlerts])
              .then(([{data: {result: nodeFSAvail}}, {data: {result: nodeFSSizeAvail}}, alerts]) => getNodePartitionsTableData(nodeFSAvail, nodeFSSizeAvail, alerts))
    , [instanceIP]),
  );
```
it depends if in the futur querying those informations independently might makes sense and then we would want to separate the calls here to benefit of react query caching

JBWatenbergScality commented on ui/src/services/prometheus/fetchMetrics.js:
```diff
@@ -0,0 +1,21 @@
+import { queryPrometheus } from './api';
+import { PORT_NODE_EXPORTER } from '../../constants';
+
+export function queryNodeFSAvail(instanceIP: string): Promise<Object> {
```
You may want to try typing prometheus answer in a more precise way based on https://prometheus.io/docs/prometheus/latest/querying/api/

ChengYanJin replied:
Yes! This is what I wanted to do to continue your typing work!

JBWatenbergScality reviewed (commented):
Hello, 
Good job for introducing react query and enhanced typing !
I've added a few comment about typing which can be a bit improved, also we have to check if we can merge https://github.com/scality/metalk8s/pull/3002 before this PR in order to benefit from https://github.com/scality/metalk8s/commit/eea1b97b0afb024657e2203a1ebbdca9f31f77fc and `useTypedSelector`.
I think we may have to review a bit the style implementation because it seems not to be adapted for other screen resolution, eg on my screen : 
![image](https://user-images.githubusercontent.com/75977494/104894107-91ac5200-5974-11eb-9cac-4b698889c82f.png)
The tbody height goes beyond my screen height and trigger the visibility of the scrollbar despite that there is no scroll necessity. I think that we should not define the tbody height and instead let the array fills the necessary place according to the number of partitions. Doing so also simplify reuse of the table component as it won't impact the style of the page where it is included.

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/2932-display-system-partitions-in-node-page` has diverged from
`development/2.8` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.8` into `improvement/2932-display-system-partitions-in-node-page`
* Rebase `improvement/2932-display-system-partitions-in-node-page` onto `origin/development/2.8`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

ChengYanJin committed:
ui/alerts: Fix the alert display issue

- No alert message display because of lacking `message` field in the
annotations, so we need to retrieve the alert from `description`
- Retrieve the alert message from annotations.summary should come from
  the last as a backup of, because it includes the minimal information.

Refs: #2932

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/2932-display-system-partitions-in-node-page` and the
destination branch `development/2.8`.

Please resolve the conflict on **the feature branch** (`improvement/2932-display-system-partitions-in-node-page`).

```bash
 $ git fetch
 $ git checkout origin/improvement/2932-display-system-partitions-in-node-page
 $ git merge origin/development/2.8
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/2932-display-system-partitions-in-node-page
```

JBWatenbergScality commented on ui/src/services/prometheus/fetchMetrics.js:
```diff
@@ -2,7 +2,15 @@
 import { PORT_NODE_EXPORTER } from '../../constants';
 
-export function queryNodeFSUsage(instanceIP: string): Promise<Object> {
+export type Prometheus = {
+  status: 'success' | 'error',
+  data: any, // diff between `Expression queries` and `Querying metadatas`
+  error: ?any,
+  errorType: ?any,
+  warnings: ?Array<String>,
+};
```
I think a more precise type would be :

```typescript
type RangeMatrixResult = {
    resultType: "matrix",
    result: {
        metric: {
            [labelName: string]: string
        },
        values: [unixTime: number, sampleValue: string][]
    }[] 
}

type InstantVectorResult = {
    resultType: "vector",
    result: {
        metric: {
            [labelName: string]: string
        },
        value: [unixTime: number, sampleValue: string]
    }[]
}

type ScalarResult = {
    resultType: "scalar",
    result: [unixTime: number, scalarValue: string]
}

type StringResult = {
    resultType: "string",
    result: [unixTime: number, stringValue: string]
}

type PrometheusQueryResult = {
    status: "error",
    errorType: string,
    error: string
} | {
    status: "success",
    data: RangeMatrixResult | InstantVectorResult | ScalarResult | StringResult,
    warnings?: string[]
}
```

ChengYanJin replied:
Thank you! 
I am wondering should we name the item of the `result.values[]`, or the following definition is enough?
```
type RangeMatrixResult = {
  resultType: 'matrix',
  result: {
    metric: {
      [labelName: string]: string,
    },
    values: [number, string][],
  }[],
};
```

JBWatenbergScality replied:
I think it is enough, no need to name the items.

ChengYanJin committed:
ui/style: Fix the scroll issue and partition table separation

Refs: #2932

ChengYanJin committed:
ui/typing: Use useTypedSelector to access the typed redux-store

Refs: #2932

ChengYanJin committed:
ui/fs: Improve the NodePartitionTable

- Fetch the data for the table together by using promise.all
- Display the loader when the status is loading
- UseTheme hook to retrieve the theme from styled-component

Resf: #2932

ChengYanJin committed:
ui/typing: Add the typing for Volume Status Prop

Refs: #2932

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

JBWatenbergScality commented on ui/src/components/NodePartitionTable.js:
```diff
@@ -92,32 +92,37 @@ const columns = [
-      alertsNodeFS.data,
-    );
+  if (status === 'success') {
+    data = partitions;
```
I think this can be dropped and you can use partitions directly

JBWatenbergScality commented on ui/src/components/NodePartitionTable.js:
```diff
@@ -151,7 +156,7 @@ const NodePartitionTable = ({ instanceIP }: { instanceIP: string }) => {
         </thead>
         <Body {...getTableBodyProps()}>
-          {data.length === 0 && (
+          {status === 'success' && data.length === 0 && (
```
We should probably display a message in case of failure

ChengYanJin committed:
ui/style: Extract the tabs items style

Refs: #2932

ChengYanJin committed:
ui/fs: Add the error condition for fetching the data

Refs: #2932

ChengYanJin committed:
ui/style: Move the commonlayoutstyle to the style folder

Refs: #2932

ChengYanJin committed:
ui/style: Replace the overflow: scroll with auto

To solve the scroll issue

Refs: #2932

ChengYanJin committed:
ui/style: Move the style for the metric action to the commonLayoutStyle

Refs: #2932

JBWatenbergScality commented on ui/src/components/NodePartitionTable.js:
```diff
@@ -91,9 +91,13 @@ const columns = [
 const NodePartitionTable = ({ instanceIP }: { instanceIP: string }) => {
   const theme = useTheme();
-  const { data: partitions, status } = useQuery(
+  let { data: partitions, status } = useQuery(
```
As we don't want to mutate partitions and status you can use `const` here

JBWatenbergScality commented on ui/src/services/NodeVolumesUtils.js:
```diff
@@ -419,3 +422,60 @@ const getPartitionSize = (
+  let removedWarningAlerts = alerts;
+  removeIndex.forEach((index) => removedWarningAlerts.splice(index, 1));
+
+  return removedWarningAlerts;
```
In order to make it easier to take new alert rules into account I think it would be preferable to split the function similarly to this approach below : 
 
```js
function getAlertUniqueKey(alert: PrometheusAlert): string | null {
     if(alert.labels.alertname.startsWith("NodeFilesystem")) {
              return alert.labels.alertname + alert. labels.device + alert.labels.instance;
    }
    if(alert.labels.alertname.startsWith("PersistentVolume")) {
             return alert.labels.alertname + alert.persistentVolumeClaim;
    }
    return null;
}

export const removeWarningAlerts = (
  alerts: Array<PrometheusAlert>,
): Array<PrometheusAlert> => {
    const unfilteredAlerts: PrometheusAlert[] = [];
    const indexedAlerts: {[key: string]: PrometheusAlert} = {}; 
    
    alerts.forEach(alert => {
        const alertUniqueKey = getAlertUniqueKey(alert);
        if  (!alertUniqueKey){
            unfilteredAlerts.push(alert);
        } else if (alertUniqueKey && ((indexedAlerts[alertUniqueKey] && indexedAlerts[alertUniqueKey].labels.severity === 'warning') || !indexedAlerts[alertUniqueKey])) {
           indexedAlerts[alertUniqueKey] = alert;
        }
    })
    
    return [...unfilteredAlerts, ...Object.values(indexedAlerts)];
}
```

JBWatenbergScality commented on ui/src/services/NodeVolumesUtils.test.js:
```diff
@@ -667,3 +668,168 @@ it('should volume09 when the index is 10', () => {
+  },
+];
+
+it('should remove the wanring alerts', () => {
```
small typo here `warning`

JBWatenbergScality commented on ui/src/services/NodeVolumesUtils.test.js:
```diff
@@ -667,3 +668,168 @@ it('should volume09 when the index is 10', () => {
+
+it('should remove the wanring alerts', () => {
+  const result = removeDiskWarningAlerts(alerts);
+  expect(result.length).toEqual(3);
```
You may want to test if result array is the expecting one using 
```js
expect(result).toStrictEqual([
  {
    annotations: {
      description:
        'Filesystem on /dev/vdc2 at 192.168.1.6:9100 has only 0.00% available space left.',
      runbook_url:
        'https://github.com/kubernetes-monitoring/kubernetes-mixin/tree/master/runbook.md#alert-name-nodefilesystemalmostoutofspace',
      summary: 'Filesystem has less than 3% space left.',
    },
    endsAt: '2021-01-21T09:32:35.358Z',
    fingerprint: '5ef4c93e954ac2c9',
    receivers: [
      {
        name: 'null',
      },
    ],
    startsAt: '2021-01-18T16:43:35.358Z',
    status: {
      inhibitedBy: [],
      silencedBy: [],
      state: 'active',
    },
    updatedAt: '2021-01-21T09:28:40.099Z',
    generatorURL:
      'http://prometheus-operator-prometheus.metalk8s-monitoring:9090/graph?g0.expr=%28node_filesystem_avail_bytes%7Bfstype%21%3D%22%22%2Cjob%3D%22node-exporter%22%7D+%2F+node_filesystem_size_bytes%7Bfstype%21%3D%22%22%2Cjob%3D%22node-exporter%22%7D+%2A+100+%3C+3+and+node_filesystem_readonly%7Bfstype%21%3D%22%22%2Cjob%3D%22node-exporter%22%7D+%3D%3D+0%29&g0.tab=1',
    labels: {
      alertname: 'NodeFilesystemAlmostOutOfSpace',
      container: 'node-exporter',
      device: '/dev/vdc2',
      endpoint: 'metrics',
      fstype: 'ext4',
      instance: '192.168.1.6:9100',
      job: 'node-exporter',
      mountpoint: '/mnt/testpart2',
      namespace: 'metalk8s-monitoring',
      pod: 'prometheus-operator-prometheus-node-exporter-m9qcj',
      prometheus: 'metalk8s-monitoring/prometheus-operator-prometheus',
      service: 'prometheus-operator-prometheus-node-exporter',
      severity: 'critical',
    },
  },
  {
    annotations: {
      description:
        'Filesystem on /dev/vdc2 at 192.168.1.6:9100 has only 0.00% available space left.',
      runbook_url:
        'https://github.com/kubernetes-monitoring/kubernetes-mixin/tree/master/runbook.md#alert-name-nodefilesystemalmostoutofspace',
      summary: 'Filesystem has less than 5% space left.',
    },
    endsAt: '2021-01-21T09:32:35.358Z',
    fingerprint: '6331188d982cdcfc',
    receivers: [
      {
        name: 'null',
      },
    ],
    startsAt: '2021-01-18T16:43:35.358Z',
    status: {
      inhibitedBy: [],
      silencedBy: [],
      state: 'active',
    },
    updatedAt: '2021-01-21T09:28:40.092Z',
    generatorURL:
      'http://prometheus-operator-prometheus.metalk8s-monitoring:9090/graph?g0.expr=%28node_filesystem_avail_bytes%7Bfstype%21%3D%22%22%2Cjob%3D%22node-exporter%22%7D+%2F+node_filesystem_size_bytes%7Bfstype%21%3D%22%22%2Cjob%3D%22node-exporter%22%7D+%2A+100+%3C+5+and+node_filesystem_readonly%7Bfstype%21%3D%22%22%2Cjob%3D%22node-exporter%22%7D+%3D%3D+0%29&g0.tab=1',
    labels: {
      alertname: 'NodeFilesystemAlmostOutOfSpace',
      container: 'node-exporter',
      device: '/dev/vdc2',
      endpoint: 'metrics',
      fstype: 'ext4',
      instance: '192.168.1.6:9100',
      job: 'node-exporter',
      mountpoint: '/mnt/testpart2',
      namespace: 'metalk8s-monitoring',
      pod: 'prometheus-operator-prometheus-node-exporter-m9qcj',
      prometheus: 'metalk8s-monitoring/prometheus-operator-prometheus',
      service: 'prometheus-operator-prometheus-node-exporter',
      severity: 'warning',
    },
  },
  {
    annotations: {
      description:
        'Filesystem on /dev/vdc1 at 192.168.1.6:9100 has only 3.28% available space left.',
      runbook_url:
        'https://github.com/kubernetes-monitoring/kubernetes-mixin/tree/master/runbook.md#alert-name-nodefilesystemalmostoutofspace',
      summary: 'Filesystem has less than 5% space left.',
    },
    endsAt: '2021-01-21T09:32:35.358Z',
    fingerprint: '6c9be09b96993ce4',
    receivers: [
      {
        name: 'null',
      },
    ],
    startsAt: '2021-01-18T16:43:35.358Z',
    status: {
      inhibitedBy: [],
      silencedBy: [],
      state: 'active',
    },
    updatedAt: '2021-01-21T09:28:40.092Z',
    generatorURL:
      'http://prometheus-operator-prometheus.metalk8s-monitoring:9090/graph?g0.expr=%28node_filesystem_avail_bytes%7Bfstype%21%3D%22%22%2Cjob%3D%22node-exporter%22%7D+%2F+node_filesystem_size_bytes%7Bfstype%21%3D%22%22%2Cjob%3D%22node-exporter%22%7D+%2A+100+%3C+5+and+node_filesystem_readonly%7Bfstype%21%3D%22%22%2Cjob%3D%22node-exporter%22%7D+%3D%3D+0%29&g0.tab=1',
    labels: {
      alertname: 'NodeFilesystemAlmostOutOfSpace',
      container: 'node-exporter',
      device: '/dev/vdc1',
      endpoint: 'metrics',
      fstype: 'xfs',
      instance: '192.168.1.6:9100',
      job: 'node-exporter',
      mountpoint: '/mnt/testpart1',
      namespace: 'metalk8s-monitoring',
      pod: 'prometheus-operator-prometheus-node-exporter-m9qcj',
      prometheus: 'metalk8s-monitoring/prometheus-operator-prometheus',
      service: 'prometheus-operator-prometheus-node-exporter',
      severity: 'warning',
    },
  },
])
```

JBWatenbergScality replied:
additionally your test seems to lack coverage on `persistentvolume` alerts and alerts of other kinds that should not be filtered

JBWatenbergScality commented on ui/src/components/NodePartitionTable.js:
```diff
@@ -0,0 +1,243 @@
+  display: flex;
+  justify-content: center;
+`;
+const NodePartitionTable = ({ instanceIP }: { instanceIP: string }) => {
```
There are some tests missing for this component. IMO we should test success, loading and error variants of this component by mocking api calls.

ChengYanJin replied:
Will add the test for this component in the coming PR.

JBWatenbergScality reviewed (changes requested):
The change-log entry referring this change seems to be missing

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @jbwatenbergscality

ChengYanJin committed:
changelog: Mention system partition tab added

Refs: #2932

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/2932-display-system-partitions-in-node-page` and the
destination branch `development/2.8`.

Please resolve the conflict on **the feature branch** (`improvement/2932-display-system-partitions-in-node-page`).

```bash
 $ git fetch
 $ git checkout origin/improvement/2932-display-system-partitions-in-node-page
 $ git merge origin/development/2.8
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/2932-display-system-partitions-in-node-page
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @jbwatenbergscality

ChengYanJin committed:
ui/test: Add the test for nodepartition table using RTL

Refs: #2932

ChengYanJin committed:
ui/component: Add aria-label for the status component

Refs: #2932

ChengYanJin committed:
ui/test: Fix the issue of NodeVolumeUtils

Refs: #2932

ChengYanJin committed:
ui/component: Add aria-label in NodePartitionTable component

Refs: #2932

ChengYanJin committed:
ui/service/apiclient: Add comment for ApiClient

TODO:
We should throw error immediately rather than catch the error and return

This may change the behavior of the error handling of the APIs.

Refs: #2932

ChengYanJin committed:
ui/components: Throw error in NodePartitionTable

Refs: #2932

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/2932-display-system-partitions-in-node-page` and the
destination branch `development/2.8`.

Please resolve the conflict on **the feature branch** (`improvement/2932-display-system-partitions-in-node-page`).

```bash
 $ git fetch
 $ git checkout origin/improvement/2932-display-system-partitions-in-node-page
 $ git merge origin/development/2.8
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/2932-display-system-partitions-in-node-page
```

ChengYanJin committed:
ui/typing: Remove the never used code thanks to typing

Refs: #2932

JBWatenbergScality commented on ui/src/components/NodePartitionTable.test.js:
```diff
@@ -0,0 +1,150 @@
+  await waitForLoadingToFinish();
+
+  // Verify
+  expect(screen.getByLabelText('status warning')).toBeInTheDocument();
+  expect(screen.getByLabelText('percentage')).toBeInTheDocument();
+});
+
+test('handles server error', async () => {
+  // S
+  jest.useFakeTimers();
+  initializeProm('http://192.168.1.18:8443/api/prometheus');
+  initializeAM('http://192.168.1.18:8443/api/alertmanager');
+  // override the default route with error status
+  server.use(
+    rest.get(
+      'http://192.168.1.18:8443/api/prometheus/api/v1/query',
+      (req, res, ctx) => {
+        return res(ctx.status(500));
+      },
+    ),
+    rest.get(
+      'http://192.168.1.18:8443/api/alertmanager/api/v2/alerts',
+      (req, res, ctx) => {
+        return res(ctx.status(500));
+      },
+    ),
+  );
+  const { getByLabelText } = render(
+    <NodePartitionTable instanceIP={'192.168.1.29'} />,
+  );
+  expect(getByLabelText('loading')).toBeInTheDocument();
+
+  // E
+  await waitForLoadingToFinish();
+
+  // V
+  expect(
+    screen.getByText('System partitions request has failed.'),
+  ).toBeInTheDocument();
+});
```
You may want to move this in a describe block to explicitly limit the scope of `beforeAll`, `afterEach` and `afterAll` functions to those tests

JBWatenbergScality commented on ui/src/components/NodePartitionTable.test.js:
```diff
@@ -0,0 +1,150 @@
+
+  // Verify
+  expect(screen.getByLabelText('status warning')).toBeInTheDocument();
+  expect(screen.getByLabelText('percentage')).toBeInTheDocument();
```
You may also want to ensure `/mnt/testpart` to be in the document, as well as used percentage (96.9 or 97 if rounded) and global size

JBWatenbergScality commented on ui/src/components/NodePartitionTable.js:
```diff
@@ -225,6 +226,7 @@ const NodePartitionTable = ({ instanceIP }: { instanceIP: string }) => {
                           percentage={cell.value}
                           buildinLabel={`${cell.value}%`}
                           backgroundColor={theme.brand.primaryDark1}
+                          aria-label={'percentage'}
```
ProgressBar component should have `${cell.value}%` as a readable text

JBWatenbergScality commented on ui/src/components/NodePartitionTable.js:
```diff
@@ -118,12 +118,17 @@ const NodePartitionTable = ({ instanceIP }: { instanceIP: string }) => {
               nodeFSSizeResult.data.result,
               alerts,
             );
+          } else if (
+            nodeFSUsageResult.error ||
+            nodeFSSizeResult.error ||
+            alerts.error
+          ) {
+            throw new Error();
```
This should be part of those functions : `queryNodeFSUsage`, `queryNodeFSSize`, `getAlerts`, if getAlerts is used by another component you may want to implement a getNodePartitionAlert that handle this

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @jbwatenbergscality

ChengYanJin committed:
ui/test: Add describe block for the tests to limit the scope of function

To make sure the functions beforeAll, afterEach, afterAll function in
the correct place.

Refs: #2932

ChengYanJin committed:
ui: Throw out the error to trigger Promise reject

Refs: #2932

ChengYanJin committed:
ui/utils: Round the number instead of pareseInt

To make the data a little more accurate

Refs: #2932

JBWatenbergScality committed:
ui: Fix kubernetes build issue

When installing kubernetes client, the library is being built in the context of the consumer project meaning that it has access to the consumer node_modules.
That bring the issue that typescript was trying to build metalk8s types dependency to build kubernetes client leading to a build failure.

The issue has been resolved by this commit : scality/kubernetes-client-javascript@579d06628b6c15648551eab591e80b0bbd7990f3 which lists explicitly the types dependency of kubernetes client.

JBWatenbergScality reviewed (approved):
Congrats, LGTM !

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

ChengYanJin commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.8`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.8`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`



Please check the status of the associated issue None.

Goodbye chengyanjin.
